src/hooks/usePlayground.js

```diff
@@ -0,0 +1,28 @@
+import { useEffect, useState } from 'react';
+import { initialValues } from '../constants';
+import state from '../lib/state';
+import { useAppContext } from '../components/Context';
+import parser from '../parser';
+const savedState = state.load();
+
+function usePlayground() {
+  const [markup, setMarkup] = useState(savedState.markup || initialValues.html);
+  const [query, setQuery] = useState(savedState.query || initialValues.js);
+  const { setParsed, htmlRoot } = useAppContext();
+
+  useEffect(() => {
+    if (!htmlRoot) {
+      return;
+    }
+
+    const parsed = parser.parse({ htmlRoot, query });
+    setParsed(parsed);
+
+    state.save({ markup: markup, query: query });
+    state.updateTitle(parsed.expression?.expression);
```

Comment:
`state.save` saves state to the window global (URL). I don't think that belongs to this hook abstraction. It will conflict if you would use the `usePlayground` hook multiple times.

Reply:
we can create an `useSave` hook and when we'll decide to save state on a server we could change the behavior of the hook. What do you think ? I have only a doubt on how to make the two hook in comunication:

1) At boot the state should be saved with the initial values 
2) On markup/query change the state should be saved again

I didn't want to add too code in `Embedded.js` and `Playground.js`

Reply:
Not sure if we need a hook for it. It's possible, but is it worth it?

> I didn't want to add too code in Embedded.js and Playground.js

Some duplicate code is fine. We can't abstract everything away. (aha programming?)

So for the state, we still need a `useEffect` in both playground and embedded. It might be possible to wrap that away, but I'm not sure if it's worth it.


```js
const [query, setQuery, markup, setMarkup] = usePlayground(state.load());
useEffect(() => state.save({ markup, query }), [markup, query]);
```

This would also leave the option open to add something like a `save button`:

```jsx
<button onClick={() => state.save({ markup, query })} />
```

Reply:
so I can do in this way

```diff 

diff --git a/src/components/Embedded.js b/src/components/Embedded.js
index e816e22..50e5b97 100644
--- a/src/components/Embedded.js
+++ b/src/components/Embedded.js
@@ -3,6 +3,8 @@ import { useLocation } from 'react-router-dom';
 import queryString from 'query-string';
 
 import usePlayground from '../hooks/usePlayground';
+import { initialValues } from '../constants';
+import state from '../lib/state';
 
 import Preview from './Preview';
 import Query from './Query';
@@ -24,9 +26,17 @@ const SUPPORTED_PANES = {
   result: true,
 };
 
+const savedState = state.load();
+
 // TODO: we should support readonly mode
 function Embedded() {
-  const [query, setQuery, markup, setMarkup] = usePlayground();
+  const [query, setQuery, markup, setMarkup] = usePlayground({
+    initialMarkup: initialValues.html || savedState.markup,
+    initialQuery: initialValues.js || savedState.js,
+  });
+  useEffect(() => {
+    state.save({ markup, query });
+  }, [markup, query]);
 
   const location = useLocation();
   const params = queryString.parse(location.search);
diff --git a/src/components/Playground.js b/src/components/Playground.js
index 20c05c3..57b42ca 100644
--- a/src/components/Playground.js
+++ b/src/components/Playground.js
@@ -1,14 +1,24 @@
-import React from 'react';
+import React, { useEffect } from 'react';
 
 import usePlayground from '../hooks/usePlayground';
+import state from '../lib/state';
+import { initialValues } from '../constants';
 
 import Preview from './Preview';
 import MarkupEditor from './MarkupEditor';
 import Result from './Result';
 import Query from './Query';
```

What do you think?

Reply:
Looks great!